go/tendermint/apps/staking/staking.go

```diff
@@ -542,63 +712,73 @@ func (app *stakingApplication) reclaimEscrow(ctx *abci.Context, state *MutableSt
 		return staking.ErrInvalidSignature
 	}
 
+	// No sense if there is nothing to reclaim.
+	if reclaim.Shares.IsZero() {
+		return staking.ErrInvalidArgument
+	}
+
+	// Verify delegator account nonce.
 	id := signedReclaim.Signature.PublicKey
-	from := state.account(id)
-	if from.Nonce != reclaim.Nonce {
+	to := state.account(id)
+	if to.General.Nonce != reclaim.Nonce {
 		app.logger.Error("ReclaimEscrow: invalid account nonce",
-			"from", id,
-			"account_nonce", from.Nonce,
+			"to", id,
+			"account_nonce", to.General.Nonce,
 			"reclaim_nonce", reclaim.Nonce,
 		)
 		return staking.ErrInvalidNonce
 	}
 
-	debondingInterval, err := state.debondingInterval()
-	if err != nil {
-		app.logger.Error("ReclaimEscrow: failed to query debonding interval",
-			"err", err,
-		)
-		return err
-	}
-	var (
-		now      = uint64(ctx.Now().Unix())
-		debondAt = from.DebondStartTime + debondingInterval
-	)
-	if now < debondAt {
-		app.logger.Error("ReclaimEscrow: in debonding interval",
-			"from", id,
-			"now", now,
-			"debond_at", debondAt,
+	// Fetch escrow account.
+	from := state.account(reclaim.Account)
+
+	// Fetch delegation.
+	delegation := state.delegation(id, reclaim.Account)
+
+	// Update delegated shares.
+	if err := delegation.Shares.Sub(&reclaim.Shares); err != nil {
```

Comment:
Hm, so "reclaim all my shares" requires querying the current share count?  We could use `SubUpTo` to make this easier, but not sure if being convenient is better than being explicit (No opinion either way).